repository/databases/oracle/block.go

```diff
@@ -0,0 +1,440 @@
+package oracle
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/blcvn/corev4-explorer/appconfig"
+	"github.com/blcvn/corev4-explorer/entities"
+	"github.com/blcvn/corev4-explorer/types"
+	"github.com/godror/godror"
+)
+
+const (
+	sql_INIT_TABLE_BLOCK = `CREATE TABLE %s (
+		"BLOCK_NUM" NUMBER,
+		"DATA_HASH" VARCHAR2(255),
+		"PRE_HASH" VARCHAR2(255),
+		"TX_COUNT" NUMBER,
+		"BLOCK_TIME" TIMESTAMP,
+		"PREV_BLOCK_HASH" VARCHAR2(255),
+		"BLOCK_HASH" VARCHAR2(255),
+		"CHANNEL_GENESIS_HASH" VARCHAR2(255),
+		"BLOCKSIZE" NUMBER,
+		"NETWORK_NAME" VARCHAR2(255) NOT NULL,
+		"CHANNEL_NAME" VARCHAR2(255) NOT NULL
+	) PARTITION BY LIST(NETWORK_NAME, CHANNEL_NAME) (%s)`
+
+	sql_INIT_INDEXES_BLOCK = `CREATE UNIQUE INDEX IDX_%s ON %s.%s (BLOCK_NUM, NETWORK_NAME, CHANNEL_NAME)`
+
+	sql_INIT_TABLE_RAW_TRANSACTION = `CREATE TABLE %s (
+		"PAYLOAD" BLOB,
+		"TX_HASH" VARCHAR2(255),
+		"BLOCK_TIME" TIMESTAMP,
+		"CHAINCODE_NAME" VARCHAR2(255),
+		"VALIDATION_CODE" VARCHAR2(255),
+		"CHAINCODE_PROPOSAL_INPUT" VARCHAR2(255),
+		"BLOCK_NUM" NUMBER,
+		"NETWORK_NAME" VARCHAR2(255),
+		"CHANNEL_NAME" VARCHAR2(255)
+	) PARTITION BY LIST(NETWORK_NAME, CHANNEL_NAME) (%s)`
+
+	sql_INIT_INDEXES_RAW_TRANSACTION = `CREATE UNIQUE INDEX IDX_%s ON %s.%s (TX_HASH, BLOCK_NUM, NETWORK_NAME, CHANNEL_NAME)`
+
+	sql_INSERT_BLOCK = `INSERT INTO %s (BLOCK_NUM, DATA_HASH, PRE_HASH, TX_COUNT, BLOCK_TIME, BLOCK_HASH, CHANNEL_GENESIS_HASH,
+		BLOCKSIZE, NETWORK_NAME, CHANNEL_NAME)
+	 VALUES (:1, :2, :3, :4, :5, :6, :7, :8, :9, :10)`
+
+	sql_INSERT_RAW_TRANSACTION = `INSERT INTO %s (
+			"TX_HASH", "BLOCK_TIME", "CHAINCODE_NAME", "VALIDATION_CODE",
+			 "CHAINCODE_PROPOSAL_INPUT", "BLOCK_NUM", "NETWORK_NAME", "CHANNEL_NAME", "PAYLOAD"
+		) VALUES (:1, :2, :3, :4, :5, :6, :7, :8, :9)`
+
+	sql_QUERY_RAW_TXS_BY_TRANSFORM = `SELECT b.BLOCK_NUM, b.TX_HASH, b.PAYLOAD, 
+		b.CHAINCODE_NAME, b.VALIDATION_CODE, 
+		b.CHANNEL_NAME, b.NETWORK_NAME, b.BLOCK_TIME, b.CHAINCODE_PROPOSAL_INPUT 
+		FROM %s a LEFT JOIN %s b 
+		ON a.BLOCK_NUM = b.BLOCK_NUM AND a.NETWORK_NAME = b.NETWORK_NAME  AND a.CHANNEL_NAME = b.CHANNEL_NAME
+		WHERE a.TASK_TRANSFORM_ID = :1 AND a."TYPE" = :2`
+)
+
+func (odb *OracleDB) InitTableBlock(ctx context.Context, networkChannelPairs [][]string) error {
```

Comment:
Add a validation check at the beginning of the `InitTableBlock` function to ensure that `networkChannelPairs` is not empty. If it is empty, return an appropriate error to avoid executing an invalid SQL query. [important]